pkg/cmd/experimental/syncgroups/ad/ldapinterface.go

```diff
@@ -0,0 +1,147 @@
+package ad
+
+import (
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/go-ldap/ldap"
+
+	"github.com/openshift/origin/pkg/auth/ldaputil"
+	ldapinterfaces "github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+)
+
+// NewLDAPInterface builds a new LDAPInterface using a schema-appropriate config
+func NewLDAPInterface(clientConfig ldaputil.LDAPClientConfig,
+	userQuery ldaputil.LDAPQueryOnAttribute,
+	groupMembershipAttributes []string,
+	userNameAttributes []string) LDAPInterface {
+	return LDAPInterface{
+		clientConfig:              clientConfig,
+		userQuery:                 userQuery,
+		userNameAttributes:        userNameAttributes,
+		groupMembershipAttributes: groupMembershipAttributes,
+		cachedUsers:               map[string]*ldap.Entry{},
+		ldapGroupToLDAPMembers:    map[string][]*ldap.Entry{},
+	}
+}
+
+// LDAPInterface extracts the member list of an LDAP group entry from an LDAP server
+// with first-class LDAP entries for user only. The LDAPInterface is *NOT* thread-safe.
+// The LDAPInterface satisfies:
+// - LDAPMemberExtractor
+// - LDAPGroupLister
+type LDAPInterface struct {
+	// clientConfig holds LDAP connection information
+	clientConfig ldaputil.LDAPClientConfig
+
+	// userQuery holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server
+	userQuery ldaputil.LDAPQueryOnAttribute
+	// groupMembershipAttributes defines which attributes on an LDAP user entry will be interpreted as its ldapGroupUID
+	groupMembershipAttributes []string
+	// UserNameAttributes defines which attributes on an LDAP user entry will be interpreted as its name
+	userNameAttributes []string
+
+	cachePopulated bool
+	// cachedUsers holds the result of user queries for later reference, indexed on user UID
+	// e.g. this will map an LDAP user UID to the LDAP entry returned from the query made using it
+	cachedUsers map[string]*ldap.Entry
+
+	ldapGroupToLDAPMembers map[string][]*ldap.Entry
+}
+
+var _ ldapinterfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ ldapinterfaces.LDAPGroupLister = &LDAPInterface{}
+
+// ExtractMembers returns the LDAP member entries for a group specified with a ldapGroupUID
+func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) ([]*ldap.Entry, error) {
+	if !e.cachePopulated {
```

Comment:
> I know you said this was the first round impl, do we want to always query cache first and fallback to a real query if not found rather than flagging?

Ok, I'll build the cache as I go.  I think the net effect is the same.